src/corecel/io/FileOrConsole.cc

```diff
@@ -0,0 +1,115 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/io/FileOrConsole.cc
+//---------------------------------------------------------------------------//
+#include "FileOrConsole.hh"
+
+#include <filesystem>
+#include <fstream>
+#include <string>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+#include "detail/UniqueFileNamer.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Open a file if it does not already exist
+std::ofstream open_file_noreplace(std::string const& filename)
+{
+    std::ofstream ofs;
+    auto flags = std::ios::out;
+#if __cplusplus >= 202302L
+    // C++23: use the noreplace flag to fail if the file exists
+    flags |= std::ios::noreplace;
+#else
+    // C++17/C++20: check non-atomically if the file exists
+    if (std::filesystem::exists(filename))
+    {
+        // Do not open
+        return ofs;
+    }
+#endif
+    ofs.open(filename, flags);
```

Comment:
```suggestion
#if __cplusplus >= 202302L
    ofs.open(filename, flags);
#else
    flags |= std::ios::ate;
    ofs.open(filename, flags);
    auto pos = ofs.tellp()
    if (pos != 0) // In case of rare race condition.
       return std::ofstream{};
#endif
```

Reply:
Clever! I guess it only catches the race condition of someone already flushing to disk so the file system can read it: it seems quite possible that multiple could open it but one might not write yet.

Reply:
Indeed, it is not foolproof but sounded simple enough while reducing somewhat the probability of trouble.  We could go further (but still not perfect) by immediately writing and flushing some marker content. To get to 'perfection' we would probably need to use some type of full fledge '[file lock' mechanism](https://github.com/boostorg/interprocess/blob/develop/include/boost/interprocess/sync/file_lock.hpp)' (See link for an example).
But this sounds overboard for code that disappear as soon as C++20.

Reply:
😄 We'd probably have to hook into the OS. Anyway, from what I've seen with typical id and file access patterns I'm skeptical this suggestion would reduce the race conditions by any appreciable amount... usually the time between querying the os and opening the file would be much less than the time for someone else to flush anything to it? This just seems like overkill.

Reply:
It is likely overkill/not-that-useful especially since it goes away after we move to C++23.

>  than the time for someone else to flush anything to it?

It it not about the user doing something explicitly but about this same code running at the same time from multiple threads or processes on the same machine (or not).  For example 2 processes starting at the same time are not unlikely to get here at the 'same' time.

Reply:
I was personifying the other threads 😅